src/H5Dmpio.c

```diff
@@ -5578,274 +5420,146 @@ H5D__mpio_get_chunk_insert_info_types(MPI_Datatype *contig_type, hbool_t *contig
 } /* end H5D__mpio_get_chunk_insert_info_types() */
 
 /*-------------------------------------------------------------------------
- * Function:    H5D__mpio_collective_filtered_io_type
+ * Function:    H5D__mpio_collective_filtered_vec_io
  *
- * Purpose:     Constructs a MPI derived datatype for both the memory and
- *              the file for a collective I/O operation on filtered chunks.
- *              The datatype contains the chunk offsets and lengths in the
- *              file and the locations of the chunk data buffers to read
- *              into/write from.
+ * Purpose:     Given a pointer to a H5D_filtered_collective_io_info_t
+ *              structure with information about collective filtered chunk
+ *              I/O, populates I/O vectors and performs vector I/O on those
+ *              chunks.
  *
  * Return:      Non-negative on success/Negative on failure
  *
  *-------------------------------------------------------------------------
  */
 static herr_t
-H5D__mpio_collective_filtered_io_type(H5D_filtered_collective_io_info_t *chunk_list, size_t num_entries,
-                                      H5D_io_op_type_t op_type, MPI_Datatype *new_mem_type,
-                                      hbool_t *mem_type_derived, MPI_Datatype *new_file_type,
-                                      hbool_t *file_type_derived)
+H5D__mpio_collective_filtered_vec_io(const H5D_filtered_collective_io_info_t *chunk_list, H5F_shared_t *f_sh,
+                                     H5D_io_op_type_t op_type)
 {
-    MPI_Aint *io_buf_array      = NULL; /* Relative displacements of filtered chunk data buffers */
-    MPI_Aint *file_offset_array = NULL; /* Chunk offsets in the file */
-    int      *length_array      = NULL; /* Filtered Chunk lengths */
-    int       mpi_code;
-    herr_t    ret_value = SUCCEED;
+    const void **io_wbufs = NULL;
+    void       **io_rbufs = NULL;
+    H5FD_mem_t   io_types[2];
+    uint32_t     iovec_count = 0;
+    haddr_t     *io_addrs    = NULL;
+    size_t      *io_sizes    = NULL;
+    herr_t       ret_value   = SUCCEED;
 
     FUNC_ENTER_PACKAGE
 
-    assert(chunk_list || 0 == num_entries);
-    assert(new_mem_type);
-    assert(mem_type_derived);
-    assert(new_file_type);
-    assert(file_type_derived);
+    assert(chunk_list);
+    assert(f_sh);
 
-    *mem_type_derived  = FALSE;
-    *file_type_derived = FALSE;
-    *new_mem_type      = MPI_BYTE;
-    *new_file_type     = MPI_BYTE;
+    if (op_type == H5D_IO_OP_WRITE)
+        iovec_count = (uint32_t)chunk_list->num_chunk_infos;
```

Comment:
So if two processes write to a chunk, after chunks are assigned, the one that doesn't own it deletes the chunk from its chunk list?

Reply:
That's correct.